### PR TITLE
[ feature ] Implement `-Werror` (WarningsAsErrors)

### PR DESCRIPTION
**Pull request by fabianhjr**

### CODE REVIEW

File: src/Idris/CommandLine.idr [hunk 1]
```diff
@@ -205,7 +207,9 @@ options = [MkOpt ["--check", "-c"] [] [CheckOnly]
               (Just "Generate profile data when compiling, if supported"),
 
            optSeparator,
-           MkOpt ["--no-shadowing-warning"] [] [IgnoreShadowingWarnings]
+           MkOpt ["-Werror"] [] [WarningsAsErrors]
+              (Just "Treat warnings as errors"),
+           MkOpt ["-Wno-shadowing"] [] [IgnoreShadowingWarnings]
```
Comment (fabianhjr) [inline src/Idris/CommandLine.idr:212]: renamed this flag for consistency (where -Wshadowing would be to enable shadow warnings in the future)

Comment (gallais) [reply src/Idris/CommandLine.idr:212]: You'll have to update the `.ipkg` files for `base` & so on then.

File: src/Idris/REPL/Common.idr [hunk 2]
```diff
@@ -108,6 +101,27 @@ emitError : {auto c : Ref Ctxt Defs} ->
...
+export
+emitWarning : {auto c : Ref Ctxt Defs} ->
+              {auto o : Ref ROpts REPLOpts} ->
+              {auto s : Ref Syn SyntaxInfo} ->
+              Warning -> Core ()
+emitWarning w
+    = do session <- getSession
+         if (session.warningsAsErrors)
+           then emitError $ WarningAsError w
+           else emitProblem w displayWarning pwarning getWarningLoc
```
Comment (gallais) [inline src/Idris/REPL/Common.idr:123]: This duplicates the logic already in `recordWarning`.

**Comment by fabianhjr** [comment]
Thanks for the feedback gallais, applied both suggestions.

File: tests/idris2/error009/expected [hunk 3]
```diff
@@ -1,5 +1,11 @@
 Error: Module DoesntExist not found
 
+(toplevel):1:1--1:1
+ 1 | import DoesntExist
+     
+
+Error: Module DoesntExist not found
+
 Exists.idr:1:1--1:19
```
Comment (fabianhjr) [inline tests/idris2/error009/expected:9]: TODO: Fix this duplicate error emission change.

File: src/Idris/CommandLine.idr (hunk 1 above)
Comment (fabianhjr) [reply src/Idris/CommandLine.idr:212]: Renamed all usages on this repo

File: src/Idris/ModTree.idr [hunk 4]
```diff
@@ -167,33 +168,36 @@ buildMod loc num len mod
...
-                                   traverse_ emitError errs
-                                   pure (ferrs ++ errs)
-                   emitWarnings
-                   traverse_ emitError ferrs
-                   pure ferrs
-           else do emitWarnings
-                   traverse_ emitError ferrs
-                   pure ferrs
+        errs <- if (not needsBuilding) then pure [] else
+           do let pad = (minus `on` length) (show len) (show num)
```
Comment (gallais) [inline src/Idris/ModTree.idr:185]:
```suggestion
           do let pad = minus (length $ show len) (length $ show num)
```

**Pull request merged**